threedi_api_client/aio/aiohttp_retry.py

```diff
@@ -0,0 +1,390 @@
+    ):
+        self.timeouts = timeouts
+        super().__init__(len(timeouts), statuses, exceptions)
+
+    def get_timeout(self, attempt: int) -> float:
+        """timeouts from a defined list."""
+        return self.timeouts[attempt]
+
+
+class _RequestContext:
+    def __init__(
+        self,
+        request: Callable[..., Any],  # Request operation, like POST or GET
+        method: str,
+        urls: Tuple[StrOrURL, ...],
+        logger: _Logger,
+        retry_options: RetryOptionsBase,
+        raise_for_status: bool = False,
+        **kwargs: Any
+    ) -> None:
+        self._request = request
+        self._method = method
+        self._urls = urls
+        self._logger = logger
+        self._retry_options = retry_options
+        self._kwargs = kwargs
+        self._trace_request_ctx = kwargs.pop('trace_request_ctx', {})
+        self._raise_for_status = raise_for_status
+
+        self._response: Optional[ClientResponse] = None
+
+    def _is_status_code_ok(self, code: int) -> bool:
+        return code not in self._retry_options.statuses and code < 500
+
+    async def _do_request(self) -> ClientResponse:
+        current_attempt = 0
+        while True:
+            self._logger.debug("Attempt {} out of {}".format(current_attempt, self._retry_options.attempts))
+            if current_attempt > 0:
+                retry_wait = self._retry_options.get_timeout(current_attempt)
+                await asyncio.sleep(retry_wait)
+
+            current_attempt += 1
+            try:
+                response: ClientResponse = await self._request(
+                    self._method,
+                    self._urls[current_attempt - 1],
+                    **self._kwargs,
+                    trace_request_ctx={
+                        'current_attempt': current_attempt,
+                        **self._trace_request_ctx,
+                    },
+                )
+            except Exception as e:
+                if current_attempt < self._retry_options.attempts:
+                    is_exc_valid = any([isinstance(e, exc) for exc in self._retry_options.exceptions])
+                    if is_exc_valid:
+                        continue
+
+                raise e
+
+            if self._is_status_code_ok(response.status) or current_attempt == self._retry_options.attempts:
+                if self._raise_for_status:
+                    response.raise_for_status()
+                self._response = response
+                return response
+
+    def __await__(self) -> Generator[Any, None, ClientResponse]:
+        return self.__aenter__().__await__()
+
+    async def __aenter__(self) -> ClientResponse:
+        return await self._do_request()
+
+    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        if self._response is not None:
+            if not self._response.closed:
+                self._response.close()
+
+
+def _url_to_urls(url: _URL_TYPE, attempts: int) -> Tuple[StrOrURL, ...]:
```

Comment:
Why do you need to create a list of urls from a url?

Reply:
No idea, this is just a copy of aiohttp_retry